From c6663850316785c43fb6f5c5cfaccf37c47a24d1 Mon Sep 17 00:00:00 2001
From: blackrebel <72fabio1@virgilio.it>
Date: Thu, 13 Apr 2017 14:20:27 +0200
Subject: [PATCH 17/27] Add more symbols in MediatekHacks

Change-Id: I64c2d76706ba387d7c018c4a9672b14b3f7d49d1
---
 libutils/MediatekHacks.cpp | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/libutils/MediatekHacks.cpp b/libutils/MediatekHacks.cpp
index bc115531..51128507 100644
--- a/libutils/MediatekHacks.cpp
+++ b/libutils/MediatekHacks.cpp
@@ -1,6 +1,22 @@
 extern "C" {
     void _ZN7android6Tracer4initEv(){}
+    void _ZNK7android9CallStack4dumpEPKc(){}
     volatile int _ZN7android6Tracer8sIsReadyE = 1;
     unsigned long long _ZN7android6Tracer12sEnabledTagsE = 0;
     int _ZN7android6Tracer8sTraceFDE = -1;
+void _ZN7android9CallStackC1EPKci(char const*, int);
+ void _ZN7android9CallStack6updateEii(int, int);
+
+ void _ZN7android9CallStackC1EPKcii(char const* logtag, int ignoreDepth, int maxDepth){
+  maxDepth = maxDepth-1;
+  maxDepth = maxDepth+1;
+  _ZN7android9CallStackC1EPKci(logtag, ignoreDepth);
+  
+ }
+
+ void _ZN7android9CallStack6updateEiii(int ignoreDepth, int maxDepth, int tid){
+  maxDepth = maxDepth-1;
+  maxDepth = maxDepth+1; 
+  _ZN7android9CallStack6updateEii(ignoreDepth, tid);
+ }
 }
-- 
2.15.1

